src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferenceCache.cs

```diff
@@ -19,36 +19,27 @@
 
 namespace Microsoft.CodeAnalysis.FindSymbols
 {
-    // TODO : this can be all moved down to compiler side.
     internal static class FindReferenceCache
     {
-        private static readonly ReaderWriterLockSlim s_gate = new();
-        private static readonly Dictionary<SemanticModel, Entry> s_cache = new();
+        private static readonly ConditionalWeakTable<SemanticModel, Entry> s_cache = new();
```

Comment:
no need for complex locking and refcounts.  it's just a CWT from semantic-model to entry.  Note: the SemanticModel is created once and lives for hte lifetime we're analyzing a document.  The FAR engine ensures we only process a document at most exactly once (for semantics) in the FAR engine.